keps/sig-cli/5295-kyaml/README.md

```diff
@@ -0,0 +1,1349 @@
+
+At the top of each KYAML document (or multi-document) will be the following header:
+
+```
+#!yaml
+---
+```
+
+This disambiguates it from JSON, since both KYAML and JSON begin with an
+opening curly-brace.  Kubernetes versions 1.33 and higher can handle KYAML
+without the header, but it is needed for versions before that.
+
+### Per-type rendering
+
+Integers and floats are rendered as their natural numeric representation.
+
+Simple strings are rendered as double-quoted and escaped. Multi-line strings
+require special consideration, see below.
+
+Booleans are rendered as `true` or `false`.
+
+Structs are rendered with curly-braces, and each field on a line of its own.
+Field names are not quoted unless they are not obviously strings (which
+includes prefixed label keys) or are one of the known type-ambiguous words
+(e.g. `no`).  Field values are rendered as per their type.  Embedded structs
+will be rendered as inline fields.
+
+KYAML does not guarantee that fields will be rendered in the order they are
+defined in the struct.  Implementations may preserve field order, but `kubectl`
+itself does not guarantee this (it converts to a map internally, which loses
+the type schema), so the order may be different.
+
+Fields which have a `json` struct tag will be rendered according to that tag:
+  * the tag’s field name will be used
+  * if the tag’s name is “-”, the field will be skipped
+  * if the tag specifies “omitempty” it will be respected
+  * if the tag specifies “omitzero” it will be respected
+
+The `yaml` struct tag will be ignored.
+
+<<[UNRESOLVED yaml self-marshaling]>>
+Should we respect `yaml` struct tags?  It is non-standard and we do
+not use it in Kubernetes.  Maybe just the name part of it?  Or the
+JSON-identical parts?
+<<[/UNRESOLVED>>
+
+Maps are rendered with curly-braces, and each key on a line of its own. Keys
+are sorted and are not quoted unless they are type-ambiguous (e.g. `no`), and
+values are rendered as per their type.
+
+<<[UNRESOLVED quoted strings]>>
+Should we treat strings for field names and map keys:
+As described above, with label keys quoted but field names not
+Allow unquoted label keys and more complex (but not ambiguous) values
+Always quote strings, including ALL map keys and field names
+<<[/UNRESOLVED>>
+
+Slices and arrays are rendered with square-braces. Elements are rendered as per
+this list.
+
+Pointers are rendered as `null` or dereferenced and rendered as per the
+pointee’s type.
+
+Types which implement the Go standard `json.Marshaler` interface
+(`MarshalJSON()`) are rendered by marshaling to JSON, unmarshaling into an
+`any`, and then rendering that `any` as KYAML.
+
+Types which implement the `yaml.Marshaler` interface (`MarshalYAML()`) are
+rendered as per their type.  `MarshalYAML()` is not a Go standard and is not
+used by KYAML.
+
+<<[UNRESOLVED yaml self-marshaling]>>
+Should we do something for YAML self-marshalling?  It is non-standard and we do
+not use it in Kubernetes.
```

Comment:
I don't think we should diverge from `-o yaml`

Reply:
Any objections from others?